### PR TITLE
feat:cw code adjustment for membership card

### PR DESCRIPTION
**Pull request by ShenDing1125**
專為CwData的memberData，擴增會員卡訊息:
1.卡名 (title)
2.ID (id)
3.到期日 (ended_at)
4.開通日 (delivered_at)

### CODE REVIEW

**Comment by ShenDing1125** [comment]
[(2)【前台】親子訂閱制：埋碼調整](https://trello.com/c/ZXa1rHBG/5296-2%E3%80%90%E5%89%8D%E5%8F%B0%E3%80%91%E8%A6%AA%E5%AD%90%E8%A8%82%E9%96%B1%E5%88%B6%EF%BC%9A%E5%9F%8B%E7%A2%BC%E8%AA%BF%E6%95%B4)

File: src/hooks/tracking.ts [hunk 1]
```diff
@@ -1069,3 +1096,62 @@ export const useTracking = (trackingOptions = { separator: '|' }) => {
...
+        filteredAndUniqueData.push({
+          ...cardEnrollment,
+          member: {
+            ...cardEnrollment.member,
+            order_logs: filteredOrderLogs,
+          },
+        })
+        cardIds.add(cardEnrollment.card_id)
+      }
+    })
```
[the remark this answers is not in the document]
Comment (ShenDing1125) [reply src/hooks/tracking.ts]: recording:
會員卡有兩種時間形式
一個是發卡時才定義起訖時間 (order_product started_at ended_at)
一個是會員卡本身就有 staryed_at ended_at

File: src/hooks/tracking.ts [hunk 2]
```diff
@@ -1069,3 +1096,71 @@ export const useTracking = (trackingOptions = { separator: '|' }) => {
     },
   }
 }
+
+export const useMemberShipCardDetails = (memberId: string | undefined = '') => {
+  if (memberId == null) memberId = ''
```
Comment (Ponchimeow) [inline src/hooks/tracking.ts:1101]: avoid using `==`

could modify at line 1123, `memberId: memberId ?? ''`

Comment (ShenDing1125) [reply src/hooks/tracking.ts]: Oh! Dealing with undefined values twice

The expression 'memberId == null' is equivalent to 'memberId === null | memberId === undefined'.
Okey,I'll avoid it in the future.

This is a good suggestion for use  "memberId: memberId ?? ''

**Review by Ponchimeow — APPROVED** [review]

File: src/hooks/tracking.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (kkshyu) [reply src/hooks/tracking.ts]: it is better to separate from uniq and filter part.
```ts
// uniq
 const cardIds = new Set()
  ;(memberShipCardDetails?.card_enrollment as CwCardEnrollment)?.forEach(cardEnrollment => {
      cardIds.add(cardEnrollment.card_id)
  })

// filter
```

**Pull request merged**